### PR TITLE
Remove code duplication in client

### PR DESCRIPTION
**Pull request by jl-wynen**
There currently is a lot of duplicate code in the client. And a lot of methods are mostly copies of each other. And other are implemented slightly differently even though they do equivalent work. This makes it hard to see similarities.

This PR unifies all interaction with the backend. This is mainly a refactor, but there are some behavioural changes:
- `datasets_create` now only returns the PID like all other creation functions.
- get-type functions only return `None` if the response is of type 'unknown id' and raise an exception otherwise (e.g. connection failure). Please check that the regular expression in `_call_endpoint` is correct!
- All operations now log after the operation was successful but not when they raise an exception. Could easily change that to also log before raising if you want that.
- Type hints are now correct
- Non-optional arguments no longer have default values.

This is not just about cleaning up code. It will also help us test downstream code by creating a fake client. This is easier with the refactor because `_call_endpoint` and `_send_to_scicat` operate on structured data and not preformatted URLs and dictionaries. I will explain this in a separate issue.

Unfortunately, test coverage is quite sparse, so I am not positive that all refactorings are correct. Please review carefully!

Question:
`datasets_datablock_create` uses 'origdatablocks'. Should this be 'datablocks' instead?

### CODE REVIEW

File: pyscicat/client.py [hunk 1]
```diff
@@ -89,68 +87,47 @@ def __init__(
...
-    #         "updatedAt": datetime.isoformat(datetime.utcnow()) + "Z"
-    #     }
-    #     sample_url = f'{self._base_url}Samples'
-
-    #     resp = self._send_to_scicat(sample_url, sample)
-    #     if not resp.ok:  # can happen if sample id is a duplicate, but we can't tell that from the response
-    #         err = resp.json()["error"]
-    #         raise ScicatCommError(f"Error creating Sample {err}")
+        return requests.request(
+            method=cmd,
```
Comment (nitrosx) [inline pyscicat/client.py:94]: WIll it be better if we validate cmd to be one of the accepted commands, aka GET, POST, etc...?

**Review by nitrosx — APPROVED** [review]

File: pyscicat/client.py (hunk 1 above)
Comment (jl-wynen) [reply pyscicat/client.py:94]: Why? The callers of this function determine what is supported, not `_send_to_scicat`.

Comment (nitrosx) [reply pyscicat/client.py:94]: Sorry, I did not see that this function is for internal use only.
In that case, I assume that the maintainers knows what they are doing.

**Comment by dylanmcreynolds** [comment]
Thanks for this PR.  This is a lot of changes, but it appears pretty clean.

The change to the urllib causes a couple of issues. flake8 is unhappy:

```
./pyscicat/client.py:734:9: F821 undefined name 'urllib'
./pyscicat/client.py:750:9: F821 undefined name 'urllib'
```

If these same problems also cause pytest to fail.

Can you please look at this and get both flake8 and pytest running?

**Comment by jl-wynen** [comment]
Done.

**Pull request merged**